src/linear_algebra/free_module.lean

```diff
@@ -117,6 +117,14 @@ instance self : module.free R R := of_basis $ basis.singleton unit R
 instance of_subsingleton [subsingleton N] : module.free R N :=
 of_basis (basis.empty N : basis pempty R N)
 
+instance direct_sum {ι : Type*} {M : ι → Type*} [Π (i : ι), add_comm_monoid (M i)]
+  [Π (i : ι), module R (M i)] [Π (i : ι), module.free R (M i)] : module.free R (⨁ i, M i) :=
+of_basis $ dfinsupp.basis $ λ i, choose_basis R (M i)
```

Comment:
Let's make the `M` explicit, to match with the other instances.

Why not (also) make an instance for `dfinsupp`?

```suggestion
instance dfinsupp {ι : Type*} (M : ι → Type*) [Π (i : ι), add_comm_monoid (M i)]
  [Π (i : ι), module R (M i)] [Π (i : ι), module.free R (M i)] : module.free R (Π₀ i, M i) :=
of_basis $ dfinsupp.basis $ λ i, choose_basis R (M i)

instance direct_sum {ι : Type*} (M : ι → Type*) [Π (i : ι), add_comm_monoid (M i)]
  [Π (i : ι), module R (M i)] [Π (i : ι), module.free R (M i)] : module.free R (⨁ i, M i) :=
module.free.dfinsupp R M
```